proposal/first-class-calc.md

```diff
@@ -0,0 +1,541 @@
+calc entered a set of functions that only expected numbers and ended up
+producing an error far away in space or time from the actual source of the
+issue. For example:
+
+* Miriam publishes a Sass library with a function, `frobnicate()`, which does a
+  bunch of arithmetic on its argument and returns a result.
+
+* Jina tries calling `frobnicate(calc(100px + 10%))`. This works, so she commits
+  it and ships to production.
+
+* Miriam updates the implementation of `frobnicate()` to call `math.log()`,
+  which does not support calcs. She doesn't realize this is a breaking change,
+  since she was only ever expecting numbers to be passed.
+
+* Jina updates to the newest version of Miriam's library and is unexpectedly
+  broken.
+
+To avoid this issue, we've made it so that the only operations that support
+calcs are those within `calc()` expressions. This follows Sass's broad principle
+of "don't design for users using upstream stylesheets in ways they weren't
+intended to be used".
+
+Going back to the example above, if Miriam *did* want to support calcs, she
+could simply wrap `calc()` around any mathematical expressions she writes. This
+will still return plain numbers when given compatible numbers as inputs, but it
+will also make it clear that `calc()`s are supported and that Miriam expects to
+support them on into the future.
+
+#### Interpolation in `calc()`
+
+Historically, interpolation has been the only means of injecting SassScript
+values into `calc()` expressions, so for backwards compatibility, we must
+continue to support it to some degree. Exactly to what degree and how it
+integrates with first-class calc is a question with multiple possible answers,
+though.
+
+The answer we settled on was to handle interpolation in a similar way to how we
+handled backwards-compatibility with Sass's [`min()` and `max()` functions]: by
+parsing `calc()` expressions using the old logic if they contain any
+interpolation and continuing to treat those values as opaque strings, and only
+using the new parsing logic for calcs that contain no interpolation. This is
+maximally backwards-compatible and it doesn't require interpolated calcs to be
+reparsed after interpolation.
+
+[`min()` and `max()` functions]: ../accepted/min-max.md
+
+Here are some alternatives we considered:
+
+1. Re-parsing a calc that contains interpolation once the interpolation has been
+   resolved, and using the result as in the calc object rather than an unquoted
+   string. For example, `calc(#{"1px + 2px"})` would return `3px` rather than
+   `calc(1px + 2px)`. However, doing another parse at evaluation-time would add
+   substantial complexity and some amount of runtime overhead. The
+   return-on-investment would also be inherently limited, since we're planning
+   on gradually transitioning users away from interpolation in `calc()` anyway.
+
+2. Treating interpolation as a `CalcValue` that participates in the normal
```

Comment:
These are formally defined below, but it could be helpful to introduce them earlier on and use the terms throughout.

Reply:
I couldn't think of a good way to summarize them here, so I added a link to their formal definitions. Let me know if you have a better idea of how to describe them.